jax/_src/third_party/scipy/linalg.py

```diff
@@ -0,0 +1,127 @@
+def _maybe_real(A, B, tol=None):
+  """
+  Return either B or the real part of B, depending on properties of A and B.
+
+  The motivation is that B has been computed as a complicated function of A,
+  and B may be perturbed by negligible imaginary components.
+  If A is real and B is complex with small imaginary components,
+  then return a real copy of B.  The assumption in that case would be that
+  the imaginary components of B are numerical artifacts.
+
+  Parameters
+  ----------
+  A : ndarray
+      Input array whose type is to be checked as real vs. complex.
+  B : ndarray
+      Array to be returned, possibly without its imaginary part.
+  tol : float
+      Absolute tolerance.
+
+  Returns
+  -------
+  out : real or complex array
+      Either the input array B or only the real part of the input array B.
+
+  """
+  if jnp.isrealobj(A) and jnp.iscomplexobj(B):
+    if tol is None:
+      tol = {0: feps*1e3, 1: eps*1e6}[_array_precision[B.dtype.char]]
+    if jnp.allclose(B.imag, 0.0, atol=tol):
+      B = B.real
+  return B
+
+def _asarray_square(A):
+  """
+  Wraps asarray with the extra requirement that the input be a square matrix.
+
+  The motivation is that the matfuncs module has real functions that have
+  been lifted to square matrix functions.
+
+  Parameters
+  ----------
+  A : array_like
+      A square matrix.
+
+  Returns
+  -------
+  out : ndarray
+      An ndarray copy or view or other representation of A.
+
+  """
+  A = jnp.asarray(A)
+  if A.ndim != 2 or A.shape[0] != A.shape[1]:
+    raise ValueError('expected square array_like input')
+  return A
+
+@jit
+def _algorithm_11_1_1(F, T):
+  # Algorithm 11.1.1 from Golub and Van Loan "Matrix Computations"
+  N = T.shape[0]
+  minden = jnp.abs(T[0, 0])
+
+  def _outer_loop(p, F_minden):
+    _, F, minden = lax.fori_loop(1, N-p+1, _inner_loop, (p, *F_minden))
+    return F, minden
+
+  def _inner_loop(i, p_F_minden):
+    p, F, minden = p_F_minden
+    j = i+p
+    s = T[i-1, j-1] * (F[j-1, j-1] - F[i-1, i-1])
+    T_row, T_col = T[i-1], T[:, j-1]
+    F_row, F_col = F[i-1], F[:, j-1]
+    ind = (jnp.arange(N) >= i) & (jnp.arange(N) < j-1)
+    val = (jnp.where(ind, T_row, 0) @ jnp.where(ind, F_col, 0) -
+            jnp.where(ind, F_row, 0) @ jnp.where(ind, T_col, 0))
+    s = s + val
+    den = T[j-1, j-1] - T[i-1, i-1]
+    s = lax.cond(den != 0, lambda s, den: s/den, lambda s, den: s, s, den)
+    F = F.at[i-1, j-1].set(s)
+    minden = jnp.minimum(minden, jnp.abs(den))
+    return p, F, minden
+
+  return lax.fori_loop(1, N, _outer_loop, (F, minden))
+
+@_wraps(scipy.linalg.funm)
+def funm(A, func, disp=True):
+  A = _asarray_square(A)
+  T, Z = schur(A)
+  T, Z = rsf2csf(T, Z)
+
+  F = jnp.diag(func(jnp.diag(T)))
+  F = F.astype(T.dtype.char)
+
+  F, minden = _algorithm_11_1_1(F, T)
+  F = Z @ F @ Z.conj().T
+  F = _maybe_real(A, F)
+
+  tol = {0: feps, 1: eps}[_array_precision[F.dtype.char]]
+
+  if minden == 0.0:
+    minden = tol
+
+  if jnp.any(jnp.isinf(F)):
```

Comment:
This is another statement that precludes transformation. Instead you can write it something like this:
```python
err = jnp.where(jnp.any(jnp.isinf(F)), jnp.inf, jnp.minimum(1, jnp.maximum(
            tol, (tol / minden) * norm(jnp.triu(T, 1), 1))))
```